h5py/tests/test_dataset.py

```diff
@@ -165,6 +166,63 @@ def test_empty_create_via_Empty_class(self):
         self.f.create_dataset('foo', data=h5py.Empty(dtype='f'))
         self.assertTrue(is_empty_dataspace(self.f['foo'].id))
 
+    def test_create_incompatible_data(self):
+        data = self.f.create_dataset('dset', (3,))
+        with self.assertRaises(ValueError):
+            self.f.create_dataset('bar', shape=4, data=data)
+
+
+class TestReadDirectly(BaseDataset):
+
+    """
+        Feature: Read data directly from Dataset into a Numpy array
+    """
+
+    def test_read_direct(self):
+        dset = self.f.create_dataset("dset", (100,), dtype='int64')
+        source_dset = self.f.create_dataset("source_dset", (100,), dtype='int64')
+        dest_dset = self.f.create_dataset("dest_dset", (100,), dtype='int64')
+        empty_dset = self.f.create_dataset("edset", dtype='int64')
+
+        source_sel = sel.select((100,), ..., source_dset)
+        dest_sel = sel.select((100,), ..., dest_dset)
```

Comment:
You don't need a `dest_dset` here - your destination is an array, so `dataset` can be None.